### DISCUSSION TITLE
Auto-healing doesn't work after some time or manual heal

### DISCUSSION THREAD

**Discussion by HaykManukyanAvetiky** [minio/minio#12672] [Q&A] [answered]
I have minio cluster which consist of 2 pools with 4 servers on each. 
``` server http://minio{1...4}/data http://minio{5...8}/data ```
I run servers with docker-compose and using ``` minio/minio:RELEASE.2021-06-17T00-10-46Z ```  image  
Each minio server/container /data dir attached to host as volume for persistent storage
```  - ./data/data6:/data ```
I am testing minio for durability and  resilience  cause I need it for petabyte scale highly available, durable and resilient workload.
Now for test purposes I was deleting   some folder/bucket in data6 directory to mimic data lose or so.

## Expected Behavior

I expect that minio will automatically heal/restore data.
## Current Behavior

It does nothing. It auto-healed several times in past but now minio did nothing , I tried to upload and download data to put some activity to trigger auto-heal but no result.  I tried ``` mc admin info ``` and it shows that everything is ok. Prometheus metrics also doesn't said anything. Even restart of container doesn't help. I just been able to recover data manually running ``` mc admin heal ``` .

## Possible Solution

The reason that I can recall for now  is that manual heal somehow turns off auto-healing, because it works before I tried manual heal very first time. I tried to find some documentation about healing of minio but it is  not enough info. When minio does/triggers auto-healing ? how/when it is checking data existence? How I can be sure the if mc admin info shows OK it is really OK ?

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Your Environment

* Version used (`minio --version`):
* Server setup and configuration:
* Operating System and version (`uname -a`):

**Comment by harshavardhana** [comment]
> Now for test purposes I was deleting some folder/bucket in data6 directory to mimic data lose or so.

Auto healing runs but it is run as per the scanner schedule you need to wait for it to figure out this information and it is not instantaneous - if you want faster scanner times look at how you reduce the scanner intervals 

```
 mc admin config set play/ scanner --env
KEY:
scanner  manage namespace scanning for usage calculation, lifecycle, healing and more

ARGS:
MINIO_SCANNER_DELAY     (float)     scanner delay multiplier, defaults to '10.0'
MINIO_SCANNER_MAX_WAIT  (duration)  maximum wait time between operations, defaults to '15s'
MINIO_SCANNER_CYCLE     (duration)  time duration between scanner cycles, defaults to '1m'
```

**Comment by klauspost** [comment]
@HaykManukyanAvetiky If you want to test a realistic scenario, replace an entire drive. Random files is very unlikely to go missing on disks, but full disk failure are common.

When the scanner run a cycle on the disk where you deleted the files it will notice the missing folders. However under normal circumstances there can be weeks or months between a bucket being visited on a specific disk. You can tune that, mostly with the scanner delay, but it will be more disruptive to your other operation.

**Comment by HaykManukyanAvetiky** [comment]
The problem is not only healing , there is no alert or indication that something happened. Also I wait more than 30m,  even after restart it is still not there, it is wired guys 

> > Now for test purposes I was deleting some folder/bucket in data6 directory to mimic data lose or so.
> 
> Auto healing runs but it is run as per the scanner schedule you need to wait for it to figure out this information and it is not instantaneous - if you want faster scanner times look at how you reduce the scanner intervals
> 
> ```
>  mc admin config set play/ scanner --env
> KEY:
> scanner  manage namespace scanning for usage calculation, lifecycle, healing and more
> 
> ARGS:
> MINIO_SCANNER_DELAY     (float)     scanner delay multiplier, defaults to '10.0'
> MINIO_SCANNER_MAX_WAIT  (duration)  maximum wait time between operations, defaults to '15s'
> MINIO_SCANNER_CYCLE     (duration)  time duration between scanner cycles, defaults to '1m'
> ```

**Comment by HaykManukyanAvetiky** [comment]
My concern is that at least i should have some indication that I have data lose , in order to run heal manually

**Comment by HaykManukyanAvetiky** [comment]
The problem with that is , when you have data lose like I said, it need to be recovered immediately otherwise some of you servers can go offline during that time and you will be unable to get object. I just thinking to manually run heal every day or so , via cron for example to ensure that, but it is not elegant solution     

> @HaykManukyanAvetiky If you want to test a realistic scenario, replace an entire drive. Random files is very unlikely to go missing on disks, but full disk failure are common.
> 
> When the scanner run a cycle on the disk where you deleted the files it will notice the missing folders. However under normal circumstances there can be weeks or months between a bucket being visited on a specific disk. You can tune that, mostly with the scanner delay, but it will be more disruptive to your other operation.

**Comment by klauspost** [comment, marked as answer]
@HaykManukyanAvetiky When a full disk is lost it will start recovery at once and run through all your files. Deleting some random files/folders does not trigger that.

If a disk is offline when data is being written it is written with additional erasure shards to compensate for the missing disk. So if you have `EC:4` and a disk is offline, the data will be written with `EC:5` effectively. If a file is read and it is found to have missing data it is queued for immediate healing. 

If the scanner observes a missing folder it will similarly check if it was deleted or needs to be restored.

Full data scans are expensive and can be disruptive to the performance of your cluster which is why "drive-by-healing" is only performed very slowly.

**Comment by HaykManukyanAvetiky** [comment]
@harshavardhana and @klauspost   Now I have got some points 
- It will recover an object when it is corrupted on GET/POST , that is tested 
- If I lose bucket folder on server/disk it won't recover it until  healing of entire namespace happen once in 30days automatically or recursive healing manually, but this case is highly unlikely. Right?
I hope that it will work even I will have really large number of objects, gt 30m

At this point I leave with following issues.
- Does Minio locks object at time of healing ?
- Is there any other way of alerting and monitoring beside Prometheus metrics connected with healing ? 
- Also as I understand manual heal with mc is obsolete and will be removed in near future so no much sense to run recursive healing of bucket on schedule yes ?

I also come across to [this post](https://github.com/minio/minio/issues/9878#issuecomment-646795214)  which is seems helpful , hope this points still valid

**Reply by harshavardhana** [reply]
> * Does Minio locks object at time of healing ?

Yes it does. 

> * Is there any other way of alerting and monitoring beside Prometheus metrics connected with healing ?

Prometheus is the way to look for healing activity.

> * Also as I understand manual heal with mc is obsolete and will be removed in near future so no much sense to run recursive healing of bucket on schedule yes ?

It is not really obsolete, it is being re-implemented

**Reply by HaykManukyanAvetiky** [reply]
> It is not really obsolete, it is being re-implemented

So if I am doing recursive healing of bucker it locks only healing object yes ? not the whole bucket

**Reply by harshavardhana** [reply]
Just the object it needs to heal.

**Reply by dumblob** [reply]
> > * Also as I understand manual heal with mc is obsolete and will be removed in near future so no much sense to run recursive healing of bucket on schedule yes ?
> 
> It is not really obsolete, it is being re-implemented

I couldn't find any more detailed schedule about how the re-implementation is going, what's missing, how it'll be tested etc. Any pointers?

Thanks!

**Reply by klauspost** [reply]
@dumblob There is no specific time frame for that.

**Comment by HaykManukyanAvetiky** [comment]
Other Question Guys,  is there any way to restore deleted object in minio ?

**Reply by harshavardhana** [reply]
Use versioning.

**Reply by HaykManukyanAvetiky** [reply]
@harshavardhana  In That case can I set up some rules or so to keep/retain only last 2 versions or last month versions , for higher storage efficiency ??? I am looking at lifecycle management but cant find exactly what i need, if you can send some guides for that that would be great

**Reply by HaykManukyanAvetiky** [reply]
@harshavardhana Ok seems I found nice [guideline ](https://docs.min.io/docs/minio-bucket-lifecycle-guide.html), pasting for others to use )) 

But Recovery of deleted file still cant find it , is there some flag or so in mc undo command to revert only deleted ones ?

**Reply by harshavardhana** [reply]
You need to use versioning there is no there way to recover deleted content, actual versions deleted cannot be recovered similar to AWS S3 style behavior.

**Reply by HaykManukyanAvetiky** [reply]
@harshavardhana  as I understand in release RELEASE.2021-07-12T02-44-53Z ordinary web UI is gone and it is redirected to console ?

**Reply by harshavardhana** [reply]
Console is embedded with MinIO @HaykManukyanAvetiky
